lib/src/vector.dart

```diff
@@ -0,0 +1,419 @@
+
+  @override
+  bool contains(Object value) => _values.contains(value);
+
+  @override
+  T elementAt(int index) {
+    if (index >= length || index < 0) {
+      throw new StateError(
+          'Index $index out of bounds for Vector of length $length');
+    }
+    return this[index];
+  }
+
+  @override
+  bool every(Predicate<T> test) => _values.every(test);
+
+  @override
+  Iterable<S> expand/*<S>*/(Iterable<S> f(T value)) => _values.expand(f);
+
+  @override
+  T firstWhere(Predicate<T> test, {Lazy<T> orElse}) =>
+      _values.firstWhere(test, orElse: orElse);
+
+  @override
+  S fold/*<S>*/(S init, S reducer(S acc, T value)) =>
+      _values.fold/*<S>*/(init, reducer);
+
+  @override
+  void forEach(void f(T value)) => _values.forEach(f);
+
+  @override
+  String join([String sep]) => _values.join(sep);
+
+  @override
+  T lastWhere(Predicate<T> test, {Lazy<T> orElse}) =>
+      _values.lastWhere(test, orElse: orElse);
+
+  @override
+  Iterable<S> map/*<S>*/(S f(T value)) => _values.map(f);
+
+  @override
+  T reduce(T f(T a, T b)) => _values.reduce(f);
+
+  @override
+  T singleWhere(Predicate<T> test) => _values.singleWhere(test);
+
+  @override
+  Iterable<T> skip(int n) => _values.skip(n);
+
+  @override
+  Iterable<T> skipWhile(Predicate<T> test) => _values.skipWhile(test);
+
+  @override
+  Iterable<T> take(int n) => _values.take(n);
+
+  @override
+  Iterable<T> takeWhile(Predicate<T> test) => _values.takeWhile(test);
+
+  @override
+  List<T> toList({bool growable: false}) =>
+      new List.from(_values, growable: growable);
+
+  @override
+  Set<T> toSet() => new Set.from(_values);
+
+  @override
+  String toString() => '[' + join(', ') + ']';
+
+  /// Creates a new Vector with the value in the given [index] replaced by
+  /// [value].
+  Vector<T> update(int index, T value);
+
+  @override
+  Iterable<T> where(Predicate<T> test) => _values.where(test);
+}
+
+class _ImmutableVector<T> extends Vector<T> {
+  @override
+  final int length;
+
+  /// Maximum depth of the tree.
+  @protected
+  final int depth;
+
+  Node<T> _root;
+
+  /// Creates a new empty Vector.
+  ///
+  /// example:
+  ///   final xs = new Vector.empty();
+  ///   print(xs);
+  ///   => []
+  _ImmutableVector.empty()
+      : length = 0,
+        depth = 1,
+        _root = new Leaf.empty(),
+        super._();
+
+  /// Private constructor for building new Vectors from path copied roots.
+  _ImmutableVector.fromRoot(this.length, this.depth, this._root) : super._();
+
+  @override
+  @protected
+  Iterable<T> get _values => _root.traverse();
+
+  @override
+  T operator [](int index) {
+    assert(index > 0);
+    if (index >= length) {
+      throw new StateError(
+          'Index out of bounds: $index is greater than $length');
+    }
+    return _root.get(index, 5 * (depth - 1));
+  }
+
+  @override
+  Vector<T> append(T value) {
+    if (length >= trieSize << (5 * (depth - 1))) {
+      final xs = new List<Node<T>>(32);
```

Comment:
I will open an issue for bench-marking so I can compare some different implementations.